### ISSUE TITLE
Give the choice to respawn dead shell at remote location

### ISSUE THREAD

**Issue by p3r7** [riscy/shx-for-emacs#16]
Hello,

First of all, thanks for this really neat package. I'm using it daily and enjoying it.

I would like to propose a customization toggle for the behavior of respawning dead shell buffers.

I can submit a PR if you want, but would first like to propose the idea.

**Current behavior**

When the comint process in interactive shell buffer dies, shx makes it come back to life whihc is really handy.

This is handled by the [first cond](https://github.com/riscy/shx-for-emacs/blob/master/shx.el#L181) of `shx-send-input`.

The only problem is that when the shell was remote, it will [respawn in a local shell instead](https://github.com/riscy/shx-for-emacs/blob/master/shx.el#L184).

**Expected behavior**

I understand that some users might want this behavior by default, but what I would personally expect is to have it respawned at the previous location (i.e. on the remote server).

I work daily with lots of remote server connections and have [helper functions](https://github.com/p3r7/prf-shell#interractive-shells-1) for connecting to them and naming the shell buffers according to the remote location.

I guess a var could be defined to let the user choose the behavior (respawn local or keep old location).

In the meantime, I have overridden `shx-send-input` in my own config to keep my sanity.

**Comment by riscy** [comment]
Yeah, this change definitely makes sense.

The reason I initially chose to restart in what I called a "safe" directory is because the reconnect blocks Emacs so, if the remote is unreachable, Emacs appears to lock up.

I think a simple solution to the blocking would be to add a message to `shx--reconnect` to tell the user what's happening and how to stop it (C-g).

**Comment by riscy** [comment]
I'm playtesting this a little (see the attached commit) and realized #ef6b7ea makes it impossible to return to a local shell once we've :ssh'd out.  I feel like that should be addressed.  A couple options:
- At some point I had it so that `:ssh` by itself returned the user to a local shell, but that almost feels like a hidden easter egg.
- We could bring up a `y-or-n-p` prompt that would ask: "Reconnect shell to remote host?"

I'm leaning toward the second option, but I thought I'd see what you thought. :)

**Comment by p3r7** [comment]
I agree with the `y-or-n-p` option.

Additionally, I think it would be nice to have a defcustom to allow the user to configure for always yes or always no.

**Comment by riscy** [comment]
This sounds good to me!  Watch for it to drop this weekend. :)

**Comment by p3r7** [comment]
Great 🙌

**Comment by p3r7** [comment]
I've been busy lately.

I'll attempt to take a look at those 2 commits this evening.

**Comment by riscy** [comment]
No worries!  No rush either.  I went with a solution that doesn't have a prompt, but I think it works.

**Comment by p3r7** [comment]
I've tested your solution and it works indeed quite smoothly.

I just have a small remark, but it might be out of the scope of `shx`.

Function `shx--validate-shell-file-name` attempts to guess which interpreter to use. It first tries value of `explicit-shell-file-name`.

One idiosyncrasy of shell.el is that when spawning an interactive shell (via command `shell`) on a remote connection, the user gets prompted for an `explicit-shell-file-name` but this value does not get stored as a (buffer-local) var.

As such, the value of command for `shx--validate-shell-file-name` will always default to `/bin/sh`, which is kinda sad (I would rather be prompted for a path than be retrograded to this old thing).

I personally deal with this annoyance by defvaring a default remote interpreter (`with-shell-interpreter-default-remote`):

```el
(defadvice shx--validate-shell-file-name (around shx--validate-shell-file-name-default-remote-interpreter activate)
    "Set `explicit-shell-file-name' to `with-shell-interpreter-default-remote' if exists"
    (let ((remote-id (file-remote-p default-directory))
          (explicit-shell-file-name explicit-shell-file-name))
      (when (and remote-id
                 (file-exists-p (concat remote-id with-shell-interpreter-default-remote)))
        (setq explicit-shell-file-name with-shell-interpreter-default-remote))
      ad-do-it))
```

But this works for me as I always use the same interpreter for remote connections.

Ideally interactive shell buffers should keep a memory of their interpreter (via a buffer-local var) so that the shell could get respawned with the exact same one. But this would be quite intrusive (relying on a wrapper around command `shell` or via an advice).

**Comment by riscy** [comment]
> Ideally interactive shell buffers should keep a memory of their interpreter (via a buffer-local var, at least their initial value) so that the shell could get respawned with the exact same one. But this would be quite intrusive (relying on a wrapper around command shell or via an advice).

It feels like it would be useful to have a customization variable that lets you map hostnames (or regexps covering different hostnames) to preferred interpreters, maybe like:
```elisp
(defcustom shx-remote-shell-commands
  '((".*.host.com" . "/bin/zsh"))
  "Mapping from remote hosts to (preferred) shell commands."
  :type '(alist :key-type regexp :value-type string))
```
and then having `shx--validate-shell-file-name` swap in the right interpreter when connecting/reconnecting.

**Comment by p3r7** [comment]
Yes, that would be even better than my solution of having a single default value.

This way one could also have a catch-all entry with `".*"`.

**Comment by p3r7** [comment]
Regarding your idea of defining an alist of host regexp / interpreter couples, I came up with this:

```el
(defun shx--keep-first (fn list)
  "Keep the first element from LIST for which FN is non-nil."
  (let (e res)
    (while (not res)
      (setq res (funcall fn (car list))
            list (cdr list)))
    res))

(defun shx--interpreter-for-host (host)
  (shx--keep-first
   (lambda (e)
     (let ((regexp (car e))
           (interp (cdr e)))
       (when
           (string-match regexp host)
         interp)))
   shx-remote-shell-commands))
```

Please note that the name of function `shx--keep-first` is inspired by function [-keep from dash.el](https://github.com/magnars/dash.el#-keep-fn-list).

I've also modified [my wrapper around shell functions](https://github.com/p3r7/prf-shell) to save `explicit-shell-file-name` as a buffer-local value.
It works but I'm not super satisfied with the implementation (super verbose, `comint` was a PITA, resetting local values).

**Comment by riscy** [comment]
I was just thinking about this!  I usually hesitate to use recursion, but another solution might be something like:

```elisp
(defun shx--interpreter-for-host (host &optional options)
  (setq options (or options shx-remote-shell-commands))
  (when options
    (let ((regexp (caar options))
          (interp (cdar options)))
      (cond ((string-match regexp host) interp)
            ((cdr options) (shx--interpreter-for-host host (cdr options)))))))
```

**Comment by p3r7** [comment]
Yes, that'd be a tad more compact.

The only drawback is a bigger stack usage (no native tail-call optimization in elisp) but given that `shx-remote-shell-commands` / `options` would never exceed a few dozen items that's a non-issue.

**Comment by p3r7** [comment]
Other than the minor comment I made regarding the regexp format, everything looks mighty fine.

I guess a mention of the feature in the README would be the last remaining thing for this ticket to be closed.

:beers:

**Comment by p3r7** [comment]
Looks fine.

Thanks for all your work!

**Comment by p3r7** [comment]
Oh, I just stumbled upon this.

It seems that we have re-implemented something that is standard since Emacs 26.

In section 6.5.2 in the [TRAMP manual](https://www.gnu.org/software/emacs/manual/html_node/tramp/Remote-processes.html), it appears that we can configure which shell to default for remote hosts.

When reading the docstring for `connection-local-criteria-alist`, it appears that we could even have a wildcard entry (`if CRITERIA is nil, it
always applies.`).

**Comment by riscy** [comment]
Ooof, good point!  I think it is worth documenting this, and then I'll probably just pull it out of shx and bump the version as a breaking change.
